File: packages/modules/api/rust/src/core/airvisual.com/airvisual.rs
use reqwest::Client;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct AirQuality {
    pub status: String,
    pub data: AirQualityData,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct AirQualityData {
    pub city: String,
    pub state: String,
    pub country: String,
    pub location: Location,
    pub current: Current,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Location {
    #[serde(rename = "type")]
    pub location_type: String,
    pub coordinates: Vec<f64>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Current {
    pub pollution: Pollution,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Pollution {
    pub aqius: i32,
    pub mainus: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct CountriesResponse {
    pub status: String,
    pub data: Vec<CountryData>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct CountryData {
    pub country: String,
}

pub struct AirVisualClient {
    pub api_key: String,
    pub base_url: String,
    pub client: Client,
}

impl AirVisualClient {
    pub fn new(api_key: String, base_url: Option<String>) -> Self {
        Self {
            api_key,
            base_url: base_url.unwrap_or_else(|| "http://api.airvisual.com/v2".to_string()),
            client: Client::new(),
        }
    }

    pub async fn get_countries(&self) -> Result<CountriesResponse, anyhow::Error> {
        let url = format!("{}/countries", self.base_url);
        let response = self
            .client
            .get(&url)
            .query(&[("key", &self.api_key)])
            .send()
            .await?
            .json::<CountriesResponse>()
            .await?;

        Ok(response)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use wiremock::matchers::{method, path, query_param};
    use wiremock::{Mock, MockServer, ResponseTemplate};

    #[tokio::test]
    async fn test_get_countries() {
        let mock_server = MockServer::start().await;

        Mock::given(method("GET"))
            .and(path("/countries"))
            .and(query_param("key", "test-key"))
            .respond_with(ResponseTemplate::new(200).set_body_json(serde_json::json!({
                "status": "success",
                "data": [{"country": "Vietnam"}]
            })))
            .mount(&mock_server)
            .await;

        let client = AirVisualClient::new("test-key".to_string(), Some(mock_server.uri()));
        let result = client.get_countries().await.unwrap();
        assert_eq!(result.status, "success");
        assert_eq!(result.data.len(), 1);
        assert_eq!(result.data[0].country, "Vietnam");
    }
}

File: packages/modules/api/go/core/airvisual.com/airvisual.go
package airvisual

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
)

var AirVisualBaseURL = "http://api.airvisual.com/v2"

type AirQuality struct {
	Status string `json:"status"`
	Data   struct {
		City     string `json:"city"`
		State    string `json:"state"`
		Country  string `json:"country"`
		Location struct {
			Type        string    `json:"type"`
			Coordinates []float64 `json:"coordinates"`
		} `json:"location"`
		Current struct {
			Pollution struct {
				Aqius  int    `json:"aqius"`
				Mainus string `json:"mainus"`
			} `json:"pollution"`
		} `json:"current"`
	} `json:"data"`
}

type CountriesResponse struct {
	Status string `json:"status"`
	Data   []struct {
		Country string `json:"country"`
	} `json:"data"`
}

func GetCountries(apiKey string) (*CountriesResponse, error) {
	u, _ := url.Parse(fmt.Sprintf("%s/countries", AirVisualBaseURL))
	q := u.Query()
	q.Set("key", apiKey)
	u.RawQuery = q.Encode()

	resp, err := http.Get(u.String())
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result CountriesResponse
	err = json.NewDecoder(resp.Body).Decode(&result)
	return &result, err
}
